### PR TITLE
Add models for generative learning and hyper parameter search

### PR DESCRIPTION
**Pull request by quantshah**
Added models for generative learning, data generation and hyper parameter search scripts.

### CODE REVIEW

File: src/qml_benchmarks/models/base.py [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+    def __init__(
+        self,
+        dim: int = None,
+        learning_rate=0.001,
+        batch_size=32,
+        max_steps=10000,
+        cdiv_steps=100,
+        convergence_interval=None,
+        random_state=42,
+        jit=True,
```
Comment (josephbowles) [inline src/qml_benchmarks/models/base.py:94]: these init arguments weren't appearing as options in my Jupyter notebook when pressing tab. Do you know if there is a way to make that work?

File: src/qml_benchmarks/models/base.py [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+                "Model not initialized. Call model.initialize first with"
+                "example data sample."
+            )
+        keys = jax.random.split(self.generate_key(), num_samples)
+
+        # we do not take the gradient wrt the sampling, so decouple the param dict here
+        params_copy = copy.deepcopy(self.params_)
+
+        for key in params_copy.keys():
+            params_copy[key] = jax.lax.stop_gradient(params_copy[key])
```
Comment (josephbowles) [inline src/qml_benchmarks/models/base.py:189]: I don't think we need to copy the params here because we aren't taking the gradient. So you can just use self.params_

File: src/qml_benchmarks/models/base.py [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+        """
+        if self.params_ is None:
+            self.initialize(X.shape[1])
+
+        c_div_loss = (
+            jax.jit(self.contrastive_divergence_loss)
+            if self.jit
+            else self.contrastive_divergence_loss
+        )
+
```
Comment (josephbowles) [inline src/qml_benchmarks/models/base.py:259]: have been thinking a bit about this. I think one option for the score that makes sense is the pseudo likelihood, which is used in the RBM I believe. 

Once could also use the maximum mean discrepancy. I'll probably add this to the package at some point. 

We can leave like this for now but we should come back to this.
